File: backend/observability.py
from __future__ import annotations

import contextvars
import asyncio
import json
import logging
import os
import re
import time
import uuid
from typing import Any
from urllib.parse import urlparse

import httpx

_request_id: contextvars.ContextVar[str] = contextvars.ContextVar("request_id", default="")
_user_id: contextvars.ContextVar[str] = contextvars.ContextVar("user_id", default="")
_class_id: contextvars.ContextVar[str] = contextvars.ContextVar("class_id", default="")
_logger = logging.getLogger("chandra")
_SENSITIVE_KEY_RE = re.compile(
    r"(authorization|token|secret|password|privatekey|private_key|apikey|api_key|content|messagecontent|messages|prompt|profile|learningprofile|filecontents|uploadedfile|extractedtext)",
    re.I,
)


def configure_logging() -> None:
    if not logging.getLogger().handlers:
        logging.basicConfig(level=logging.INFO)


def current_request_id() -> str:
    return _request_id.get()


def current_user_id() -> str:
    return _user_id.get()


def current_class_id() -> str:
    return _class_id.get()


def set_request_id(value: str) -> contextvars.Token[str]:
    return _request_id.set(safe_request_id(value) or create_request_id())


def reset_request_id(token: contextvars.Token[str]) -> None:
    _request_id.reset(token)


def set_current_user_id(value: str) -> None:
    if value:
        _user_id.set(value)


def clear_current_user_id() -> None:
    _user_id.set("")


def set_current_class_id(value: str) -> None:
    if value:
        _class_id.set(value)


def clear_current_class_id() -> None:
    _class_id.set("")


def create_request_id() -> str:
    return str(uuid.uuid4())


def safe_request_id(value: str | None) -> str:
    candidate = (value or "").strip()

    if not candidate or len(candidate) > 128:
        return ""

    if not all(character.isalnum() or character in "_.:/-" for character in candidate):
        return ""

    return candidate


def log_event(event: str, level: str = "info", **fields: Any) -> None:
    payload = redact_log_fields({
        "dt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "environment": better_stack_environment(),
        "event": event,
        "level": level,
        "message": event,
        "service": "chandra-backend",
        **{key: value for key, value in fields.items() if value not in (None, "")},
    })
    message = json.dumps(payload, default=str, separators=(",", ":"))

    if level == "error":
        _logger.error(message)
    elif level == "warning":
        _logger.warning(message)
    else:
        _logger.info(message)

    send_better_stack_log(payload)


def log_request(*, route: str, method: str, status: int, latency_ms: float) -> None:
    log_event(
        "api.request",
        route=route,
        method=method,
        status=status,
        latencyMs=round(latency_ms),
        requestId=current_request_id(),
        userId=current_user_id(),
    )


def log_provider_failure(
    *,
    provider: str,
    provider_error_class: str,
    provider_status: int | None = None,
) -> None:
    log_event(
        "provider.failure",
        level="error",
        provider=provider,
        providerErrorClass=provider_error_class,
        providerStatus=provider_status,
        requestId=current_request_id(),
        classId=current_class_id(),
        userId=current_user_id(),
    )


async def capture_exception(error: BaseException, **context: Any) -> None:
    log_event(
        context.pop("event", "error.captured"),
        level="error",
        errorClass=error.__class__.__name__,
        requestId=current_request_id(),
        userId=current_user_id(),
        **context,
    )


def redact_log_fields(fields: dict[str, Any]) -> dict[str, Any]:
    return redact_value(fields)


def better_stack_logging_status() -> dict[str, Any]:
    source_token = os.getenv("BETTER_STACK_SOURCE_TOKEN", "").strip()
    ingesting_host = os.getenv("BETTER_STACK_INGESTING_HOST", "").strip()

    return {
        "environment": better_stack_environment(),
        "ingestingHostConfigured": bool(ingesting_host),
        "sourceTokenConfigured": bool(source_token),
        "status": "ok" if source_token and ingesting_host else "missing_config",
    }


def send_better_stack_log(payload: dict[str, Any]) -> None:
    endpoint = better_stack_ingest_endpoint()
    source_token = os.getenv("BETTER_STACK_SOURCE_TOKEN", "").strip()

    if not endpoint or not source_token:
        return

    try:
        loop = asyncio.get_running_loop()
        loop.create_task(post_better_stack_log(endpoint, source_token, payload))
    except RuntimeError:
        return


async def post_better_stack_log(endpoint: str, source_token: str, payload: dict[str, Any]) -> None:
    try:
        async with httpx.AsyncClient(timeout=1.5) as client:
            await client.post(
                endpoint,
                headers={
                    "Authorization": f"Bearer {source_token}",
                    "Content-Type": "application/json",
                },
                json=payload,
            )
    except Exception:
        return


def better_stack_ingest_endpoint() -> str:
    host = os.getenv("BETTER_STACK_INGESTING_HOST", "").strip()

    if not host:
        return ""

    candidate = host if host.startswith("http") else f"https://{host}"
    parsed = urlparse(candidate)

    if not parsed.scheme or not parsed.netloc:
        return ""

    return candidate.rstrip("/")


def better_stack_environment() -> str:
    return "production" if os.getenv("BETTER_STACK_ENV", "").strip().lower() == "production" else "development"


def redact_value(value: Any) -> Any:
    if isinstance(value, list):
        return [redact_value(item) for item in value]

    if isinstance(value, dict):
        return {
            key: "[REDACTED]" if _SENSITIVE_KEY_RE.search(str(key)) else redact_value(nested_value)
            for key, nested_value in value.items()
        }

    return value

File: frontend/lib/observability.ts
type LogLevel = "debug" | "info" | "warn" | "error";

type LogValue = string | number | boolean | null | undefined | LogValue[] | { [key: string]: LogValue };
type LogFields = Record<string, LogValue>;

type CaptureExceptionContext = {
  event?: string;
  requestId?: string;
  route?: string;
  method?: string;
  userId?: string;
  provider?: string;
  providerErrorClass?: string;
  providerStatus?: number;
};

const serviceName = "chandra-frontend";
let browserMonitoringInitialized = false;
const sensitiveKeyPattern =
  /(authorization|token|secret|password|privatekey|private_key|apikey|api_key|content|messagecontent|messages|prompt|profile|learningprofile|filecontents|uploadedfile|extractedtext)/i;

export function requestIdFromRequest(request: Request): string {
  const headerValue =
    request.headers.get("x-request-id") ||
    request.headers.get("x-cloud-trace-context")?.split("/")?.[0] ||
    "";

  return safeRequestId(headerValue) || createRequestId();
}

export function createRequestId(): string {
  return globalThis.crypto?.randomUUID?.() ?? `req_${Date.now().toString(36)}_${Math.random().toString(36).slice(2)}`;
}

export function withRequestIdHeader<T extends Response>(response: T, requestId: string): T {
  response.headers.set("X-Request-Id", requestId);
  return response;
}

export function logApiRequest(fields: {
  route: string;
  method: string;
  status: number;
  latencyMs: number;
  requestId: string;
  userId?: string;
}) {
  logEvent("api.request", "info", {
    latencyMs: Math.round(fields.latencyMs),
    method: fields.method,
    requestId: fields.requestId,
    route: fields.route,
    status: fields.status,
    userId: fields.userId
  });
}

export function logProviderFailure(fields: {
  provider: string;
  providerErrorClass: string;
  providerStatus?: number;
  requestId?: string;
  route?: string;
}) {
  logEvent("provider.failure", "error", fields);
}

export function logEvent(event: string, level: LogLevel, fields: LogFields = {}) {
  const payload = redactLogFields(compactLogFields({
    dt: new Date().toISOString(),
    environment: betterStackEnvironment(),
    event,
    level,
    message: event,
    service: serviceName,
    ...fields
  }));
  const serialized = JSON.stringify(payload);

  if (level === "error") {
    console.error(serialized);
  } else if (level === "warn") {
    console.warn(serialized);
  } else {
    console.log(serialized);
  }

  sendBetterStackLog(payload);
}

export async function captureException(error: unknown, context: CaptureExceptionContext = {}) {
  const errorClass = error instanceof Error ? error.name : "NonError";

  logEvent(context.event ?? "error.captured", "error", {
    errorClass,
    method: context.method,
    provider: context.provider,
    providerErrorClass: context.providerErrorClass,
    providerStatus: context.providerStatus,
    requestId: context.requestId,
    route: context.route,
    userId: context.userId
  });
}

export function initBrowserErrorMonitoring() {
  if (browserMonitoringInitialized || typeof window === "undefined") {
    return;
  }

  browserMonitoringInitialized = true;

  window.addEventListener("error", (event) => {
    void captureException(event.error ?? event.message, {
      event: "browser.error",
      route: window.location.pathname
    });
  });

  window.addEventListener("unhandledrejection", (event) => {
    void captureException(event.reason, {
      event: "browser.unhandled_rejection",
      route: window.location.pathname
    });
  });
}

export function redactLogFields(fields: LogFields): Record<string, LogValue> {
  return redactValue(fields) as Record<string, LogValue>;
}

export function betterStackLoggingStatus() {
  const sourceToken = process.env.BETTER_STACK_SOURCE_TOKEN?.trim() ?? "";
  const ingestingHost = process.env.BETTER_STACK_INGESTING_HOST?.trim() ?? "";

  return {
    environment: betterStackEnvironment(),
    ingestingHostConfigured: Boolean(ingestingHost),
    sourceTokenConfigured: Boolean(sourceToken),
    status: sourceToken && ingestingHost ? "ok" as const : "missing_config" as const
  };
}

export function pingBetterStackHeartbeat(url: string | undefined, event: string) {
  const heartbeatUrl = url?.trim();

  if (!heartbeatUrl) {
    logEvent("better_stack.heartbeat.missing_config", "warn", { eventType: event });
    return;
  }

  try {
    void fetch(heartbeatUrl, {
      cache: "no-store",
      method: "GET",
      signal: AbortSignal.timeout(1500)
    }).catch((caughtError: unknown) => {
      logEvent("better_stack.heartbeat.failed", "warn", {
        errorClass: caughtError instanceof Error ? caughtError.name : "NonError",
        eventType: event
      });
    });
  } catch (caughtError) {
    logEvent("better_stack.heartbeat.failed", "warn", {
      errorClass: caughtError instanceof Error ? caughtError.name : "NonError",
      eventType: event
    });
  }
}

function compactLogFields(fields: LogFields): LogFields {
  return Object.fromEntries(
    Object.entries(fields).filter(([, value]) => value !== undefined && value !== "")
  ) as LogFields;
}

function safeRequestId(value: string) {
  const trimmed = value.trim();

  if (!trimmed || trimmed.length > 128 || !/^[a-zA-Z0-9_.:/-]+$/.test(trimmed)) {
    return "";
  }

  return trimmed;
}

function sendBetterStackLog(payload: Record<string, LogValue>) {
  const endpoint = betterStackIngestEndpoint();
  const sourceToken = process.env.BETTER_STACK_SOURCE_TOKEN?.trim();

  if (!endpoint || !sourceToken) {
    return;
  }

  try {
    void fetch(endpoint, {
      body: JSON.stringify(payload),
      headers: {
        Authorization: `Bearer ${sourceToken}`,
        "Content-Type": "application/json"
      },
      method: "POST"
    }).catch(() => {
      // Logging must never break request handling.
    });
  } catch {
    // Logging must never break request handling.
  }
}

function betterStackIngestEndpoint() {
  const host = process.env.BETTER_STACK_INGESTING_HOST?.trim();

  if (!host) {
    return "";
  }

  try {
    return new URL(host.startsWith("http") ? host : `https://${host}`).toString().replace(/\/$/, "");
  } catch {
    return "";
  }
}

function betterStackEnvironment() {
  const value = process.env.BETTER_STACK_ENV?.trim().toLowerCase();
  return value === "production" ? "production" : "development";
}

function redactValue(value: LogValue): LogValue {
  if (Array.isArray(value)) {
    return value.map((item) => redactValue(item));
  }

  if (!value || typeof value !== "object") {
    return value;
  }

  return Object.fromEntries(
    Object.entries(value).map(([key, nestedValue]) => [
      key,
      sensitiveKeyPattern.test(key) ? "[REDACTED]" : redactValue(nestedValue)
    ])
  ) as Record<string, LogValue>;
}
